src/frontend/src/service/api-service.ts

```diff
@@ -0,0 +1,41 @@
+import axios from 'axios';
+
+type MethodOptions =
+  | 'POST'
+  | 'post'
+  | 'GET'
+  | 'get'
+  | 'DELETE'
+  | 'delete'
+  | 'head'
+  | 'HEAD'
+  | 'put'
+  | 'PUT'
+  | 'patch'
+  | 'PATCH';
+
+type Req = {
+  url: string;
+  data?: object;
+  method: MethodOptions;
+};
+
+export default function apiService(req: Req) {
+  return axios
+    .request(req)
+    .then(res => [res.statusText, res.data, res.status])
+    .then(([statusText, data, status]) => {
+      if (statusText) {
+        return {
+          data,
+          status,
+          statusText
+        };
+      }
+    })
+    .catch(e => {
+      return {
+        error: e.message
+      };
+    });
```

Comment:
Correct me if I'm wrong-- 

but as written, all calls to the `apiService` function will return a promise that resolves to an Object, regardless of whether they succeed or fail. This means that people using `apiService` will need to check whether the returned Object has an `'error'` key or not to determine whether or not the request succeeded. 

Why shouldn't we just return the axios promise to people using `apiService`? The `.then` and `.catch` Promise APIs were created for this use case, so I think we should use them in application code. It's similar to a `try` / `catch` block where you want to separate the business logic for success and failure so people don't have to use an if statement to figure out what to do. (The other issue here is that people may want to access properties of the error other than the message, e.g. the error code.)

I'm open to iterating on this API. It doesn't have to be 100% correct the first time. Just my $0.02.

Reply:
Are you suggesting that we cut out all of the logic and leave it up to the user on how to handle the response and or error.  The only purpose of the api-service.ts file then would be to make sure that the correct format is used when sending a request through Axios by using TypeScript.  In this case the code would look as follows.
export default function apiService(request: Request) {
  return axios
    .request(request)
    .then(res => res)
}

Reply:
Not necessarily. If you want to keep the `.then` logic that's fine with me, as long as we're open to possible changes if someone needs to access a property of the result object other than `statusText`, `data` or `status`.

My main point is that I think in application code, people should be able to write (A):

    apiService(...).then(result => {

    }).catch(result => {

    })

if they prefer, rather than requiring everyone to write (B):

    apiService(...).then(result => {
      if (result.hasOwnProperty('error')) {

      } else {
      
      }
    })

If this second way were better in all cases, then why would the creators of Promises ever have added the `catch` API? There are things you can do with A that you can't do with B, and sometimes it's good to be able to do those things in application code. If you prefer B that's fine. You can add a second wrapper for apiService that's optional to use. I'm happy to talk about this more at the meeting next week.

Reply:
I removed the then logic to leave it up to the user on how they want to handle the response and or error.